### PR TITLE
Packs - Keep metadata files in index

### PR DESCRIPTION
**Pull request by liorblob**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/23155

## Description
Keep old metadata files in the index folder and update the new one in upload process

### CODE REVIEW

File: Tests/Marketplace/upload_packs.py [hunk 1]
```diff
@@ -158,10 +159,28 @@ def update_index_folder(index_folder_path, pack_name, pack_path):
         index_folder_subdirectories = [d for d in os.listdir(index_folder_path) if
                                        os.path.isdir(os.path.join(index_folder_path, d))]
         index_pack_path = os.path.join(index_folder_path, pack_name)
+        metadata_files_in_index = glob.glob(f"{os.path.join(index_folder_path, pack_name)}/**/metadata-*.json")
```
Comment (amshamah419) [inline Tests/Marketplace/upload_packs.py:162]: Why not use `index_pack_path`?

File: Tests/Marketplace/upload_packs.py [hunk 2]
```diff
@@ -158,10 +159,28 @@ def update_index_folder(index_folder_path, pack_name, pack_path):
         index_folder_subdirectories = [d for d in os.listdir(index_folder_path) if
                                        os.path.isdir(os.path.join(index_folder_path, d))]
         index_pack_path = os.path.join(index_folder_path, pack_name)
+        metadata_files_in_index = glob.glob(f"{index_pack_path}/**/metadata-*.json")
```
Comment (IkaDemisto) [inline Tests/Marketplace/upload_packs.py:162]:
```suggestion
        metadata_files_in_index = glob.glob(f"{index_pack_path}/**/metadata-*.json", recursive=True)
```
Aiiiiii, `index_pack_path` already points to pack directory inside `index` folder (you have added additional subfolder) and by default the `recursive` flag is set to False. 
Or alternatively just remove subfolder:  `glob.glob(f"{index_pack_path}/metadata-*.json")`

File: Tests/Marketplace/upload_packs.py [hunk 3]
```diff
@@ -158,10 +159,28 @@ def update_index_folder(index_folder_path, pack_name, pack_path):
...
-            shutil.rmtree(index_pack_path)
-        shutil.copytree(pack_path, index_pack_path)
+            for d in os.scandir(index_pack_path):
+                if d.path not in metadata_files_in_index:
+                    os.remove(d.path)
+
+        # Copy new files and add metadata for latest version
+        for d in os.scandir(pack_path):
+            shutil.copy(d.path, index_pack_path)
+            if pack_version and Pack.METADATA in d.name:
```
Comment (IkaDemisto) [inline Tests/Marketplace/upload_packs.py:181]:
```suggestion
            if pack_version and Pack.METADATA == d.name:
```
Mainly for readability.

File: Tests/Marketplace/upload_packs.py [hunk 4]
```diff
@@ -473,7 +492,8 @@ def main():
             pack.cleanup()
             continue
 
-        task_status = update_index_folder(index_folder_path=index_folder_path, pack_name=pack.name, pack_path=pack.path)
+        task_status = update_index_folder(index_folder_path=index_folder_path, pack_name=pack.name, pack_path=pack.path,
```
Comment (IkaDemisto) [inline Tests/Marketplace/upload_packs.py:495]: I think after all done, we can add `pack.cleanup()` after line 502 (pack.status = PackStatus.SUCCESS.name), just to clear unnecessary pack folder

File: Tests/Marketplace/upload_packs.py [hunk 5]
```diff
@@ -158,10 +159,28 @@ def update_index_folder(index_folder_path, pack_name, pack_path):
         index_folder_subdirectories = [d for d in os.listdir(index_folder_path) if
                                        os.path.isdir(os.path.join(index_folder_path, d))]
         index_pack_path = os.path.join(index_folder_path, pack_name)
+        metadata_files_in_index = glob.glob(f"{index_pack_path}/**/metadata-*.json")
+        new_metadata_path = ''
```
[the remark this answers is not in the document]
Comment (liorblob) [reply Tests/Marketplace/upload_packs.py:163]: It's a problem with private packs, we don't have the version there. What should we do there?

[the remark this answers is not in the document]
Comment (IkaDemisto) [reply Tests/Marketplace/upload_packs.py:163]: Why it is a problem? Pack version is set to empty string by default in that function. The result will be `metadata-.json` and you are using `new_metadata_path` anyway in line 182

**Comment by liorblob** [comment]
@yuvalbenshalom just need tests

**Pull request merged**